### PR TITLE
Extract Projection, Filter, Window in LogicalPlan as independent struct

### LINKED ISSUE #1302: Extract `Projection`, `Filter`, `Window` in LogicalPlan as independent struct
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
There are some discussions in https://github.com/apache/arrow-datafusion/issues/1228

**Describe the solution you'd like**
like https://github.com/apache/arrow-datafusion/pull/1290

**Describe alternatives you've considered**
The task is mechanical, however, it is ideal for newcomers to familiarize themselves with datafusion's build, test, and PR processes. 
**Additional context**
If you want to do it, please ping me and assign the task to you, avoiding collisions with other people.

### PR DESCRIPTION
**Pull request by ic4y**
Closes #1302 

# Which issue does this PR close?

Related to #1228

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by xudong963** [comment]
@ic4y Thanks for your contribution. You need to fix the lint error, after fixing it, I'll start reviewing!

**Comment by ic4y** [comment]
@xudong963 
Thank you！I have fixed the error.  please reviewing

File: ballista/rust/core/src/serde/logical_plan/to_proto.rs [hunk 1]
```diff
@@ -777,9 +778,9 @@ impl TryInto<protobuf::LogicalPlanNode> for &LogicalPlan {
                     )))
                 }
             }
-            LogicalPlan::Projection {
+            LogicalPlan::Projection(ProjectionPlan {
```
Comment (xudong963) [inline ballista/rust/core/src/serde/logical_plan/to_proto.rs:781]: I think we can use the style, `LogicalPlan::Projection(p) => {p.var1, p.var2, ...}`.

FYI, https://github.com/apache/arrow-datafusion/issues/1228#issuecomment-969443381

File: ballista/rust/core/src/serde/logical_plan/to_proto.rs [hunk 2]
```diff
@@ -794,7 +795,7 @@ impl TryInto<protobuf::LogicalPlanNode> for &LogicalPlan {
                     },
                 ))),
             }),
-            LogicalPlan::Filter { predicate, input } => {
+            LogicalPlan::Filter(FilterPlan { predicate, input }) => {
```
Comment (xudong963) [inline ballista/rust/core/src/serde/logical_plan/to_proto.rs:798]: ditto, you can also check other places.

File: datafusion/src/logical_plan/plan.rs [hunk 3]
```diff
@@ -67,41 +108,12 @@ pub enum JoinConstraint {
...
-    /// Window its input based on a set of window spec and window function (e.g. SUM or RANK)
-    Window {
-        /// The incoming logical plan
-        input: Arc<LogicalPlan>,
-        /// The window function expression
-        window_expr: Vec<Expr>,
-        /// The schema description of the window output
-        schema: DFSchemaRef,
-    },
+    /// the ProjectionPlan
```
Comment (xudong963) [inline datafusion/src/logical_plan/plan.rs:111]: I think we don't need annotation here. cc @alamb 

nit: It's good to start with an upper case, such as `The projection plan`.

File: datafusion/src/logical_plan/plan.rs [hunk 4]
```diff
@@ -67,41 +108,12 @@ pub enum JoinConstraint {
...
+    Projection(ProjectionPlan),
+    /// the FilterPlan
+    Filter(FilterPlan),
+    /// the WindowPlan
```
Comment (xudong963) [inline datafusion/src/logical_plan/plan.rs:115]: ditto

**Review by xudong963 — COMMENTED** [review]
Thanks, @ic4y, nice work.

**Comment by xudong963** [comment]
BTW, you need to solve conflicts.

**Comment by ic4y** [comment]
@xudong963 Thanks, I will do it today

**Comment by xudong963** [comment]
@ic4y  After you finish it, ping me and I'll take a look

**Comment by ic4y** [comment]
@xudong963 Thanks, I have finished.

**Comment by xudong963** [comment]
LGTM! 



> …ct (#1302)

You can edit here correctly.

File: datafusion/src/logical_plan/plan.rs [hunk 5]
```diff
@@ -57,6 +57,47 @@ pub enum JoinConstraint {
...
+pub struct FilterPlan {
+    /// The predicate expression, which must have Boolean type.
+    pub predicate: Expr,
+    /// The incoming logical plan
+    pub input: Arc<LogicalPlan>,
+}
+
+/// Window its input based on a set of window spec and window function (e.g. SUM or RANK)
+#[derive(Clone)]
+pub struct WindowPlan {
```
Comment (alamb) [inline datafusion/src/logical_plan/plan.rs:92]: Thank you @ic4y 

I feel like these names are not consistent with the other structs in this file

How about renaming 
```
ProjectionPlan --> Projection
FilterPlan --> Filter
WindowPlan --> Window
```

To be consistent with the naming of  the model of  `CrossJoin` (not `CrossJoinPlan`), `Repartition` (not `RepartitonPlan`), etc)?

File: ballista/rust/core/src/serde/logical_plan/to_proto.rs [hunk 6]
```diff
@@ -778,43 +779,43 @@ impl TryInto<protobuf::LogicalPlanNode> for &LogicalPlan {
                     )))
                 }
             }
-            LogicalPlan::Projection {
-                expr, input, alias, ..
-            } => Ok(protobuf::LogicalPlanNode {
+            LogicalPlan::Projection(projection_plan) => Ok(protobuf::LogicalPlanNode {
```
Comment (liukun4515) [inline ballista/rust/core/src/serde/logical_plan/to_proto.rs:782]: Same comment in the [pull request](https://github.com/apache/arrow-datafusion/pull/1326#discussion_r753618119)

File: datafusion/src/logical_plan/plan.rs (hunk 5 above)
Comment (ic4y) [reply datafusion/src/logical_plan/plan.rs:92]: OK, thanks @alamb  I will modify according to your suggestion.

File: ballista/rust/core/src/serde/logical_plan/to_proto.rs (hunk 6 above)
Comment (ic4y) [reply ballista/rust/core/src/serde/logical_plan/to_proto.rs]: @liukun4515 My thoughts are the same as you, @xudong963 What you think？
I submit it first, and then modify it if necessary.

Comment (liukun4515) [reply ballista/rust/core/src/serde/logical_plan/to_proto.rs]: The modification looks good to me.

**Review by houqp — APPROVED** [review]
LGTM

**Review by alamb — APPROVED** [review]
Thanks @ic4y @liukun4515  and @xudong963 . Really appreciate your help and progress here

**Pull request merged**